src/Take.Elephant.Redis/RedisSet.cs

```diff
@@ -11,53 +11,100 @@ namespace Take.Elephant.Redis
-            : this(setName, StackExchange.Redis.ConnectionMultiplexer.Connect(configuration), serializer, db, readFlags, writeFlags, useScanOnEnumeration, true)
-        { }
+        public RedisSet(string setName,
+                        string configuration,
+                        ISerializer<T> serializer,
+                        int db = 0,
+                        CommandFlags readFlags = CommandFlags.None,
+                        CommandFlags writeFlags = CommandFlags.None,
+                        bool useScanOnEnumeration = true,
+                        bool supportEmptySets = false,
+                        TimeSpan? emptyIndicatorExpiration = default)
+            : this(setName, StackExchange.Redis.ConnectionMultiplexer.Connect(configuration), serializer, db, readFlags, writeFlags, useScanOnEnumeration, true, supportEmptySets, emptyIndicatorExpiration)
+        {
+        }
 
-        public RedisSet(string setName, IConnectionMultiplexer connectionMultiplexer, ISerializer<T> serializer, int db = 0, CommandFlags readFlags = CommandFlags.None, CommandFlags writeFlags = CommandFlags.None, bool useScanOnEnumeration = true, bool disposeMultiplexer = false)
+        public RedisSet(string setName,
+                        IConnectionMultiplexer connectionMultiplexer,
+                        ISerializer<T> serializer,
+                        int db = 0,
+                        CommandFlags readFlags = CommandFlags.None,
+                        CommandFlags writeFlags = CommandFlags.None,
+                        bool useScanOnEnumeration = true,
+                        bool disposeMultiplexer = false,
+                        bool supportEmptySets = false,
+                        TimeSpan? emptyIndicatorExpiration = default)
             : base(setName, connectionMultiplexer, db, readFlags, writeFlags, disposeMultiplexer)
         {
-            if (serializer == null) throw new ArgumentNullException(nameof(serializer));
-            _serializer = serializer;
+            _serializer = serializer ?? throw new ArgumentNullException(nameof(serializer));
             _useScanOnEnumeration = useScanOnEnumeration;
+            _supportEmptySets = supportEmptySets;
+            _emptyIndicatorExpiration = emptyIndicatorExpiration;
         }
 
-        #region ISet<T> Members
-
-        public virtual Task AddAsync(T value, CancellationToken cancellationToken = default)
+        public virtual async Task AddAsync(T value, CancellationToken cancellationToken = default)
         {
-            if (value == null) throw new ArgumentNullException(nameof(value));
+            if (value == null)
+            {
+                throw new ArgumentNullException(nameof(value));
+            }
+
             var database = GetDatabase();
-            return database.SetAddAsync(Name, _serializer.Serialize(value), WriteFlags);
+            // This must be done this way (instead of awaiting each one separately)
```

Comment:
Sorry, I don't get it: GetDatabase is a protected method that does not return an `ITransaction`, so this code as is now doesn't need to be concerned about that.

What am I missing? :)

Reply:
Unfortunately, due to how `InternalSet.GetDatabase()` is implemented, the `RedisSetMap`'s underlying `RedisSet` will participate in any transaction the set map starts (practically, the `InternalSet`'s overriden `GetDatabase` impl returns the `ITransaction`). Thus, if I had used normal async-await code here, a deadlock would occur on `RedisSetMap.TryAddAsync`, since it would both await the transaction execution and the operation being executed inside the transaction, which is seemingly unsupported by StackExchange.Redis (https://stackoverflow.com/questions/25976231/stackexchange-redis-transaction-methods-freezes)